File: Complementar - Estruturas de dados (Didático)/Algoritmos19/Alg19.py
'''
♦ Algoritmo 19: Algoritmo de Pesquisa/Busca para a Estrutura de Dados Linear Dinâmica (não Estática) Lista Ligada (ou Encadeada) Simples: Busca em Lista Ligada.
'''

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.size = 0

    def insert(self, value):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node
        self.size += 1

    def remove(self, value):
        if self.head is None:
            return

        if self.head.value == value:
            self.head = self.head.next
            self.size -= 1
            return

        current = self.head
        prev = None

        while current is not None:
            if current.value == value:
                prev.next = current.next
                self.size -= 1
                return
            prev = current
            current = current.next

    def size(self):
        return self.size

    def display(self):
        current = self.head
        while current is not None:
            print(current.value, end=" ")
            current = current.next
        print()

    def search(self, value):
        current = self.head
        while current is not None:
            if current.value == value:
                return True  # Valor encontrado
            current = current.next
        return False  # Valor não encontrado

    def find_minimum(self):
        if self.head is None:
            return None

        current = self.head
        min_value = current.value

        while current is not None:
            if current.value < min_value:
                min_value = current.value
            current = current.next
        return min_value

    def find_maximum(self):
        if self.head is None:
            return None

        current = self.head
        max_value = current.value

        while current is not None:
            if current.value > max_value:
                max_value = current.value
            current = current.next
        return max_value

    def get_successor(self, value):
        current = self.head
        while current is not None:
            if current.value == value:
                if current.next is not None:
                    return current.next.value
                break  # Último elemento, não há sucessor
            current = current.next
        return None  # Valor não encontrado ou é o último

    def get_predecessor(self, value):
        if self.head is None or self.head.value == value:
            return None  # Valor não encontrado ou é o primeiro

        current = self.head
        while current is not None and current.next is not None:
            if current.next.value == value:
                return current.value
            current = current.next
        return None  # Valor não encontrado

# Exemplo de uso
list = LinkedList()
list.insert(10)
list.insert(20)
list.insert(30)

print("Tamanho da lista:", list.size())
print("Elementos da lista:", end=" ")
list.display()

print("Buscar 20:", list.search(20))
print("Menor valor:", list.find_minimum())
print("Maior valor:", list.find_maximum())
print("Sucessor de 20:", list.get_successor(20))
print("Predecessor de 20:", list.get_predecessor(20))

list.remove(20)

print("Tamanho da lista após remoção:", list.size())
print("Elementos da lista após remoção:", end=" ")
list.display()

File: Complementar - Estruturas de dados (Didático)/Algoritmos19/Alg19.js
/*
♦ Algoritmo 19: Estrutura de Dados Linear Dinâmica (não Estática) de Pesquisa/Busca para Algoritmo 12: →[Estrutura de Dados Linear Dinâmica (não Estática) de Lista Ligada/Encadeada Simples (não Duplamente Ligada) de tipo de dados genérico].
→ Operações:
    • Busca de um valor em uma Lista Ligada Simples;
    • Identificar o menor valor em uma Lista Ligada Simples;
    • Identificar o maior valor em uma Lista Ligada Simples;
    • Retornar o sucessor de um valor em uma Lista Ligada Simples;
    • Retornar o predecessor de um valor em uma Lista Ligada Simples.
*/

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.size = 0;
    }

    insert(value) {
        const newNode = new Node(value);
        newNode.next = this.head;
        this.head = newNode;
        this.size++;
    }

    remove(value) {
        if (this.head === null) {
            return;
        }

        if (this.head.value === value) {
            this.head = this.head.next;
            this.size--;
            return;
        }

        let current = this.head;
        let prev = null;

        while (current !== null) {
            if (current.value === value) {
                prev.next = current.next;
                this.size--;
                return;
            }
            prev = current;
            current = current.next;
        }
    }

    getSize() {
        return this.size;
    }

    display() {
        let current = this.head;
        while (current !== null) {
            process.stdout.write(current.value + " ");
            current = current.next;
        }
        console.log();
    }

    search(value) {
        let current = this.head;
        while (current !== null) {
            if (current.value === value) {
                return true; // Valor encontrado
            }
            current = current.next;
        }
        return false; // Valor não encontrado
    }

    findMinimum() {
        if (this.head === null) {
            return null;
        }

        let current = this.head;
        let min = current.value;

        while (current !== null) {
            if (current.value < min) {
                min = current.value;
            }
            current = current.next;
        }
        return min;
    }

    findMaximum() {
        if (this.head === null) {
            return null;
        }

        let current = this.head;
        let max = current.value;

        while (current !== null) {
            if (current.value > max) {
                max = current.value;
            }
            current = current.next;
        }
        return max;
    }

    getSuccessor(value) {
        let current = this.head;
        while (current !== null) {
            if (current.value === value) {
                if (current.next !== null) {
                    return current.next.value;
                }
                break; // Último elemento, não há sucessor
            }
            current = current.next;
        }
        return null; // Valor não encontrado ou é o último
    }

    getPredecessor(value) {
        if (this.head === null || this.head.value === value) {
            return null; // Valor não encontrado ou é o primeiro
        }

        let current = this.head;
        while (current !== null && current.next !== null) {
            if (current.next.value === value) {
                return current.value;
            }
            current = current.next;
        }
        return null; // Valor não encontrado
    }
}

const list = new LinkedList();
list.insert(10);
list.insert(20);
list.insert(30);

console.log("Tamanho da lista: " + list.getSize());
process.stdout.write("Elementos da lista: ");
list.display();

console.log("Buscar 20: " + list.search(20));
console.log("Menor valor: " + list.findMinimum());
console.log("Maior valor: " + list.findMaximum());
console.log("Sucessor de 20: " + list.getSuccessor(20));
console.log("Predecessor de 20: " + list.getPredecessor(20));

list.remove(20);

console.log("Tamanho da lista após remoção: " + list.getSize());
process.stdout.write("Elementos da lista após remoção: ");
list.display();
